server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java

```diff
@@ -24,28 +29,65 @@ public class RescorePhase {
     private RescorePhase() {}
 
     public static void execute(SearchContext context) {
-        if (context.size() == 0 || context.collapse() != null || context.rescore() == null || context.rescore().isEmpty()) {
+        if (context.size() == 0 || context.rescore() == null || context.rescore().isEmpty()) {
             return;
         }
 
         TopDocs topDocs = context.queryResult().topDocs().topDocs;
         if (topDocs.scoreDocs.length == 0) {
             return;
         }
+        TopFieldGroups topGroups = null;
+        if (topDocs instanceof TopFieldGroups topFieldGroups) {
+            assert context.collapse() != null;
+            topGroups = topFieldGroups;
+        }
         try {
             for (RescoreContext ctx : context.rescore()) {
                 topDocs = ctx.rescorer().rescore(topDocs, context.searcher(), ctx);
                 // It is the responsibility of the rescorer to sort the resulted top docs,
                 // here we only assert that this condition is met.
                 assert context.sort() == null && topDocsSortedByScore(topDocs) : "topdocs should be sorted after rescore";
             }
+            if (topGroups != null) {
+                assert context.collapse() != null;
+                /**
+                 * Since rescorers don't preserve collapsing, we must reconstruct the group and field
+                 * values from the originalTopGroups to create a new {@link TopFieldGroups} from the
+                 * rescored top documents.
+                 */
+                topDocs = rewriteTopGroups(topGroups, topDocs);
+            }
             context.queryResult()
                 .topDocs(new TopDocsAndMaxScore(topDocs, topDocs.scoreDocs[0].score), context.queryResult().sortValueFormats());
         } catch (IOException e) {
             throw new ElasticsearchException("Rescore Phase Failed", e);
         }
     }
 
+    private static TopFieldGroups rewriteTopGroups(TopFieldGroups originalTopGroups, TopDocs rescoredTopDocs) {
+        assert originalTopGroups.fields.length == 1 && SortField.FIELD_SCORE.equals(originalTopGroups.fields[0])
+            : "rescore must always sort by score descending";
+        Map<Integer, Object> docIdToGroupValue = new HashMap<>();
```

Comment:
quibble, maybe `Maps.newMapWithExpectedSize`?

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/107779/commits/c1fb5e33186a7142bd588e1345004538fc2ab98c